### PR TITLE
Makes pipenet rebuilds yield much more often

### PR DESCRIPTION
**Pull request by LemonInTheDark**
## About The Pull Request

Basically, pipenet rebuilds were blocking operations. My first commit made them sleep in between full rebuilds, but that didn't really solve the game lag that came from the operation. So I refactored how pipeline rebuilds work so I could split out each expansion step and check in to see if we're overrunning between each one.

The code is mildly pain, but I'm honestly not sure how to make it suck less. Behavior's solid tho.

Oh and a short note, some parts of the code rely on atmos components always having a parent. This wasn't always the case before, but it's more likely now. I've cleaned up what I can spot at first glance, but some more might shake out over time. I'll be keeping an eye out.

For a more indepth explanation, read this
<details>
Basically, the master controller doesn't like it when a subsystem does more work then it's allowed to, since it'll end up eating into the time we preserve for letting byond do byond things.

Because of this, we need to check somewhat often to see if we've used up our quota or not. We do that with `MC_TICK_CHECK()`
pipenet rebuilds weren't yielding at all, so they'd cause a lot of overrun, or visible lag
I tried to fix that with this commit <https://github.com/tgstation/tgstation/pull/57511/commits/debdb20526f21185152d333e3d489efeb8c2bfbd>, but I realized that the overwork wasn't just in processing the whole queue at once, it was instead mostly concentrated in this bit of code (`build_pipeline`) since it spends a lot of time expanding the edges of the pipeline
<https://github.com/tgstation/tgstation/blob/1f97c96507b9516d732ffbbbd2332d6cf4b18742/code/modules/atmospherics/machinery/datum_pipeline.dm#L35>

So I ended up moving that expansion to a subprocess in the air subsystem, and setting things up such that it checks with master each step.
This should make things that cause rebuilds produce less visible lag, and it'll very much lower the lag output of the singularity.

This approach isn't valid everywhere mind, some things need to process all at once. Excited group breakdown is a good example of this, because having a lot of gas dumped in/having a tile break partway through processing would cause shitty shitty behavior <https://github.com/tgstation/tgstation/blob/master/code/modules/atmospherics/environmental/LINDA_turf_tile.dm#L424>, but because pipelines were already kinda doing this, it's both possible and not as hellish. I needed to remove the separate var we were using for tracking volume, since that broke shit when pipelines brokedown midway through rebuilding, but outside that it seems to work like a charm
</details>

## Why It's Good For The Game

Yeets a blocking operation, things that cause pipenet rebuilds often, like explosions or the singularity will cause much less TD now.
People will bother me about atmos lagging less.

## Changelog
:cl:
refactor: Things that break pipes will cause much less game lag now. Looking at you singularity
/:cl:

### CODE REVIEW

File: code/modules/atmospherics/machinery/atmosmachinery.dm [hunk 1]
```diff
@@ -96,9 +96,9 @@
 	return
 
 /**
- * Called by all machines when on_construction() is called, it builds the network for the node
+ * Returns a list of new pipelines that need to be built up
  */
-/obj/machinery/atmospherics/proc/build_network()
+/obj/machinery/atmospherics/proc/get_rebuild_canidates()
```
Comment (Ghilker) [inline code/modules/atmospherics/machinery/atmosmachinery.dm:101]: canidates > candidates

Comment (LemonInTheDark) [reply code/modules/atmospherics/machinery/atmosmachinery.dm:101]: WHOOOP

**Comment by LemonInTheDark** [comment]
I've added a var like `turf/open/var/excited` to prevent list blaot

**Review by Mothblocks — CHANGES_REQUESTED** [review]

File: code/modules/atmospherics/machinery/datum_pipeline.dm [hunk 2]
```diff
@@ -27,12 +32,35 @@
...
 		reconcile_air()
 	update = air.react(src)
 
+///Preps a pipeline for rebuilding, insterts it into the rebuild queue
 /datum/pipeline/proc/build_pipeline(obj/machinery/atmospherics/base)
+	building = TRUE
+	var/volume = 0
+	if(!istype(base, /obj/machinery/atmospherics/pipe))
+		addMachineryMember(base)
+	else
```
[the remark this answers is not in the document]
Comment (LemonInTheDark) [reply code/modules/atmospherics/machinery/datum_pipeline.dm:48]: I think that'd make things hard to read too, but I'll do it if you insist

File: code/controllers/subsystem/air.dm [hunk 3]
```diff
@@ -485,9 +564,12 @@ GLOBAL_LIST_EMPTY(colored_images)
 
 	for(var/A in 1 to atmos_machines.len)
 		AM = atmos_machines[A]
-		AM.build_network()
+		var/list/targets = AM.get_rebuild_targets()
+		for(var/datum/pipeline/boi as anything in targets)
```
[the remark this answers is not in the document]
Comment (LemonInTheDark) [reply code/controllers/subsystem/air.dm:568]: honestly I wasn't sure what to call em. Maybe to_rebuild?

**Pull request merged**